jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java

```diff
@@ -187,12 +185,16 @@ public PendingFutureLimiter changeThresholdFactor(float thresholdFactor) {
             awaitOpportunityToEnqueueAndPurge();
         }
 
-        counter.incrementAndGet();
-        pendingFutures.put(future, System.currentTimeMillis());
-        future.handleAsync((any, exc) -> {
+        CompletableFuture<T> resultFuture = future.handleAsync((any, exc) -> {
```

Comment:
It will be inappropriate to increment this counter before the future will be enqueued. So I moved this listener's call down, after enqueuing. 

But! It changed a ThresholdListener's callback invoking. 

BEFORE my change onHiLimitReached used to be called after maxPendingCount reached, and future is coming to be enqueued. Also, amount of calling onHiLimitReached wasn't determinated - it could happen more than once if parallel enqueue has place.

AFTER my change onHiLimitReached is called exactly once when the maxPendingCount'th future is successfully enqueued. So, it will be called one future earlier than it used to.

Covered up ThresholdListener's calls behavior with couple of tests, removed enqueue_channelBlockReading_blockUnblock test as it seemed to me unclear and uncovering some cases.

Reply:
Let me clarify my remark: 
Seems like there is a case when thresholdListener.onHiLimitReached will be invoked multiple times in case of concurrent invocation of blockingEnqueue. 

In the current code counter can be decremented before increment. We even can get negative value: `resultFuture.handleAsync((any, exc) -> { ` can be executed before 
`long queueSize = counter.incrementAndGet();`
 This is error prone. 
Suggestion do increment first, decrement second.

Reply:
To avoid negative values in "counter" placed incrementing it before assigning .handleAsync.

Reply:
Hmm, maybe you did not understand the problem completely. 
Could you please take a look:
```
print("A")
resultFuture.handleAsync((any, exc) -> {
  print("B")
)
print("C")
```
In given example this is allowed:  ACB,  ABC.
If so, why you are allowing behavior when you removing resultFuture from map
and then adding it to the map?  
Could you please take a look and check if this behavior leads us to OOM ?

Reply:
You are right, I have totally missed this memory leak possibility. Fixed.